bin/add_frontend_server.rb

```diff
@@ -0,0 +1,93 @@
+  frontend_instances << frontends_for_reservation unless frontends_for_reservation.size == 0
+end
+
+instance_distribution = frontend_instances.each_with_object(Hash.new(0)){|(instance, _), instance_distribution| instance_distribution[instance.placement.availability_zone] += 1}
+
+determined_instance_zone = instance_distribution.select { |_, v| v == instance_distribution.values.min }.keys[0]
+
+puts "Using underscaled instance zone #{determined_instance_zone}"
+
+instance_name = 'frontend-' + determined_instance_zone[-1, 1] + (instance_distribution[determined_instance_zone] + 1).to_s
+puts "Naming instance #{instance_name}"
+
+run_instance_response = ec2client.run_instances ({
+                                                    dry_run: false,
+                                                    min_count: 1,
+                                                    max_count: 1,
+                                                    image_id: 'ami-d05e75b8',  #Image ID for ubuntu instance we use
+                                                    instance_type: 'c3.8xlarge',
+                                                    monitoring: {
+                                                        enabled: true
+                                                    },
+                                                    disable_api_termination: true, #Prevent against accidental termination
+                                                    placement: {
+                                                        availability_zone: determined_instance_zone
+                                                    },
+                                                    block_device_mappings: [
+                                                        {
+                                                            device_name: '/dev/sda1',
+                                                            ebs: {
+                                                                volume_size: 128,
+                                                                delete_on_termination: true,
+                                                                volume_type: 'gp2',
+                                                            },
+                                                        },
+                                                    ],
+                                                    security_groups: ['pegasus'],
+                                                })
+
+instance_id = run_instance_response.instances[0].instance_id
+puts "Looking for instance id  #{instance_id}"
+
+started_at = Time.now
+ec2client.wait_until(:instance_running, instance_ids: [instance_id]) do |waiting|
+  if Time.now - started_at > MAX_WAIT_TIME
+    puts "Instance #{instance_id} still not created. Giving up - check the EC2 console and see if there's an error."
+    exit(1);
+  end
```

Comment:
If you wanted this code to timeout properly, you need to put the check within a `before_wait` block as shown in the [documentation](http://docs.aws.amazon.com/sdkforruby/api/Aws/EC2/Client.html#wait_until-instance_method) example:
```
started_at = Time.now
client.wait_until(...) do |w|

  # disable max attempts
  w.max_attempts = nil

  # poll for 1 hour, instead of a number of attempts
  before_wait do |attempts, response|
    throw :failure if Time.now - started_at > 3600
  end

end
```